src/call/buffers.rs

```diff
@@ -0,0 +1,570 @@
+            self.buffer_offset = 0;
+        }
+
+        debug_assert!(self.buffer_offset <= buffer_len);
+        Ok(self.buffer_slice.range_from(self.buffer_offset))
+    }
+
+    fn consume(&mut self, amt: usize) {
+        self.remaining -= amt;
+        self.buffer_offset += amt;
+    }
+}
+
+#[cfg(feature = "prost-codec")]
+impl Buf for MessageReader {
+    fn remaining(&self) -> usize {
+        self.remaining
+    }
+
+    fn bytes(&self) -> &[u8] {
+        // This is similar but not identical to `BuffRead::fill_buf`, since `self`
+        // is not mutable, we can only return bytes up to the end of the current
+        // slice.
+
+        // Optimization for empty slice
+        if self.buffer_slice.len() == 0 {
+            return &[];
+        }
+
+        debug_assert!(self.buffer_offset <= self.buffer_slice.len());
+        self.buffer_slice.range_from(self.buffer_offset)
+    }
+
+    fn advance(&mut self, mut cnt: usize) {
+        // Similar but not identical to `BufRead::consume`. We must also advance
+        // the buffer slice if we have exhausted the current slice.
+
+        // The number of bytes remaining in the current slice.
+        let mut remaining = self.buffer_slice.len() - self.buffer_offset;
+        while remaining <= cnt {
+            self.consume(remaining);
+            if self.remaining == 0 {
+                return;
+            }
+
+            cnt -= remaining;
+            self.buffer_slice = self.reader.next_slice();
+            self.buffer_offset = 0;
+            remaining = self.buffer_slice.len();
+        }
+
+        self.consume(cnt);
+    }
+}
+
+/// A zero-copy writer.
+///
+/// This is implemented by writing into a Rust `Vec`, then wrapping that with the
+/// necessary gRPC data structures (see `vec_slice`).
+pub struct MessageWriter {
+    pub write_buffer: Vec<u8>,
+}
+
+impl MessageWriter {
+    /// Create an empty MessageWriter.
+    pub fn new() -> MessageWriter {
+        MessageWriter {
+            write_buffer: Vec::new(),
+        }
+    }
+
+    pub fn clear(&mut self) {
+        self.write_buffer.clear();
+    }
+
+    /// Allocates `size` bytes for writing to and returns a pointer to the start
+    /// of the newly allocated memory for writing into.
+    pub fn reserve(&mut self, size: usize) -> &mut [u8] {
+        let old_len = self.write_buffer.len();
+        let new_len = old_len + size;
+        self.write_buffer.reserve(size);
+        unsafe {
+            self.write_buffer.set_len(new_len);
+            &mut self.write_buffer[old_len..]
+        }
+    }
+
+    /// Safety: the caller takes responsibility for destroying the returned
+    /// byte_buffer. Clears the internal buffer.
+    pub unsafe fn byte_buffer(&mut self) -> *mut grpc_byte_buffer {
+        let mut vec = Vec::new();
+        mem::swap(&mut self.write_buffer, &mut vec);
+        let slice = vec_slice(vec);
+        grpc_sys::grpc_raw_byte_buffer_create(Box::into_raw(slice), 1)
```

Comment:
I think slice is leak as the pointer is ignored silently.

Reply:
In every case this is called, the pointer ends up being stored in `grpcwrap_batch_context::send_message` and is properly destroyed when the response is handled.

Reply:
`grpc_raw_byte_buffer_create` will allocate an array, and copy the slice into it. Hence, the `Box` itself will leak.

Reply:
And the leak of `Box` will always keep a ref count of the slice, so the slice will also leak.

Reply:
afaict, `grpc_raw_byte_buffer_create` doesn't allocate the *slice* only the *buffer*. The slice's ref count is incremented and the pointer (which is the box here is copied). When the buffer is destroyed then the slice is unrefed, which will call `vec_slice_unref` which will recreate the Box and drop it.